src/combinator.rs

```diff
@@ -0,0 +1,305 @@
+    /// instance of `IO`, which is performed immediately afterwards.
+    ///
+    /// A similar, albeit much less 'powerful' version of this combinator,
+    /// is [`map`][Self::map]. It does not perform this merging step, but
+    /// instead composes a function at the end of the chain.
+    pub fn bind <U, F> (self, f: F) -> IO<'a, U>
+    where
+        F: FnOnce (T) -> IO<'a, U> + 'a,
+        U: 'a,
+    {
+        // Prepare the state.
+        let IO (mut a) = self;
+
+        IO::from_fn(move |Context { i, o, e }| {
+
+            // Perform the first action in order to get the input to
+            // the function.
+            let x = a.perform(Context { i, o, e })?;
+
+            // Call the function and extract the wrapped inner value.
+            let IO (mut a) = f(x);
+
+            // Perform the returned action.
+            a.perform(Context { i, o, e })
+
+        })
+    }
+
+    /// Apply a function to an immutable reference of the output value
+    /// to produce some sort of side effect. This effect is then also
+    /// performed before the original output value is returned.
+    ///
+    /// This can be useful for debugging, for example:
+    ///
+    /// ```
+    /// use iom::{ IO, pure, debug, void };
+    ///
+    /// # fn main () -> std::io::Result<()> {
+    /// #[derive(Debug)]
+    /// enum Greeting { Hello, Meow }
+    ///
+    /// let chain = pure(Greeting::Meow)
+    ///     .tap(debug)
+    ///     .replace(Greeting::Hello)
+    ///     .tap(debug);
+    ///
+    /// let result = IO::capture(void(chain))?;
+    ///
+    /// assert_eq!(result, "Meow\nHello\n");
+    /// # Ok (()) }
+    /// ```
+    pub fn tap <F> (self, f: F) -> IO<'a, T>
+    where 
+        F: for<'b> FnOnce (&'b T) -> IO<'b> + 'a,
+    {
+        // Prepare the state.
+        let IO (mut a) = self;
+
+        IO::from_fn(move |Context { i, o, e }| {
+            let x = a.perform(Context { i, o, e })?;
+            f(&x).0.perform(Context { i, o, e })?;
+            Ok (x)
+        })
+    }
+
+    /// Like [`tap`](IO::tap), but "impure": its side effects happen
+    /// outside of the [`IO`] monad. This has some niche uses, for
+    /// example [`trace`](IO::trace) is implemented in terms of
+    /// `tap_impure`.
+    pub fn tap_impure <F> (self, f: F) -> IO<'a, T>
+    where
+        F: for<'b> FnOnce (&'b T) + 'a,
+    {
+        self.tap(|x| pure(f(x)))
+    }
+
+    /// Write `msg` to stderr *directly*, leaving the value untouched.
+    ///
+    /// The value written to stderr will *not* be captured in any way.
+    ///
+    /// ```
+    /// use iom::{ IO, empty };
+    ///
+    /// # fn main () -> std::io::Result<()> {
+    /// let chain = empty.trace("I'm").trace("spamming");
+    /// let result = IO::capture(chain)?;
+    ///
+    /// assert!(result.is_empty());
+    /// # Ok (()) }
+    /// ```
+    pub fn trace (self, msg: impl Display + 'a) -> IO<'a, T> {
```

Comment:
Should trace require impl Debug instead?

Reply:
In this case I'd say no, since the second argument to `trace` is a message that is printed to stderr, not an object that is "traced"